File: jobtracker/providers/whittard.py
import requests
from pyquery import PyQuery
from .provider import Provider
from datetime import datetime

class Whittard(Provider):
    @staticmethod
    def get_jobs():
        jobs = []
        html_page = requests.get("https://careers.whittard.co.uk/contact/").text
        requester = PyQuery(html_page)
        all_jobs = requester(".job-cta a:first")
        for job in all_jobs:
            job_details = {}
            job_url = job.attrib["href"]
            html_page = requests.get(job_url).text
            requester = PyQuery(html_page)

            #TODO: Verify each value
            job_details["link"] = job_url
            job_details["title"] = requester(".position_title .jobs-row-input")[0].text
            job_details["description"] = requester(".position_description .jobs-row-input p:first")[0].text
            job_details["type"] = requester(".position_employment_type .jobs-row-input")[0].text
            job_details["city"] = requester(".position_job_location .jobs-row-input").remove('svg').text()
            job_details["date"] = datetime.strptime(
                requester(".type-date-posted .jobs-row-input").remove('svg').text(),
                "%B %d, %Y"
            )
            
            jobs.append(job_details)

        return jobs

File: providers/whittard.go
package providers

import (
	"github.com/PuerkitoBio/goquery"
	"net/http"
)

type whittard struct{}

func NewWhittard() *whittard {
	return &whittard{}
}

func (whittard *whittard) requestJob(url string, fn func(job *Job)) error {
	res, err := http.Get(url)
	if err != nil {
		return err
	}
	defer res.Body.Close()

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return err
	}

	job := Job{
		Link:    url,
		Company: "Whittard",
	}

	doc.Find(".position_title .jobs-row-input").Each(func(i int, s *goquery.Selection) {
		job.Title = s.Text()
	})
	doc.Find(".position_description .jobs-row-input p").Each(func(i int, s *goquery.Selection) {
		job.Desc += s.Text()
	})
	doc.Find(".position_employment_type .jobs-row-input").Each(func(i int, s *goquery.Selection) {
		job.Type = s.Text()
	})
	doc.Find(".position_job_location .jobs-row-input").Each(func(i int, s *goquery.Selection) {
		job.Location = s.Text()
	})

	fn(&job)
	return nil
}

func (whittard *whittard) RetrieveJobs(fn func(job *Job)) error {
	res, err := http.Get("https://careers.whittard.co.uk/contact/")
	if err != nil {
		return err
	}
	defer res.Body.Close()

	if res.StatusCode != 200 {
		return handleStatus(res)
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return err
	}

	doc.Find(".job-cta a").Each(func(i int, s *goquery.Selection) {
		url, ok := s.Attr("href")
		if !ok {
			return
		}

		err = whittard.requestJob(url, fn)
	})
	return err
}
